crates/iceberg/src/delete_file_index.rs

```diff
@@ -199,18 +218,22 @@ pub(crate) struct DeletesForDataFile<'a> {
     state: Arc<RwLock<DeleteFileIndexState>>,
     data_file: &'a DataFile,
     seq_num: Option<i64>,
+    waker_set: Arc<WakerSet>,
 }
 
 impl Future for DeletesForDataFile<'_> {
     type Output = Result<Vec<FileScanTaskDeleteFile>>;
 
-    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match self.state.try_read() {
             Ok(guard) => match guard.deref() {
                 DeleteFileIndexState::Populated(idx) => Poll::Ready(Ok(
                     idx.get_deletes_for_data_file(self.data_file, self.seq_num)
                 )),
-                _ => Poll::Pending,
+                _ => {
+                    self.waker_set.insert(cx);
+                    Poll::Pending
+                }
             },
             Err(err) => Poll::Ready(Err(Error::new(ErrorKind::Unexpected, err.to_string()))),
         }
```

Comment:
As mentioned in #1323:
- manual `Future` implementation is error-prone. I think `async fn` should be preferred.
- Another bug here: the `try_read` may return a `TryLockError::WouldBlock`, which should not be an err. It's like a `Pending`.

Reply:
Hi @xxchan - thanks for the feedback. Good spot on the `WouldBlock` error, I've updated accordingly. Regarding the `Future` vs `async fn`, I agree that `async fn` has fewer sharp edges but conceptually in this case I think that `Future` is a better fit. Since the `Future` required in this particular case is not a particularly complex one I think it is justified. We're essentially waiting on something to be ready before we proceed with its value which is exactly the intended use case of a `Future`.

Reply:
> We're essentially waiting on something to be ready before we proceed with its value which is exactly the intended use case of a Future.

async fn is just a Future, so I think this point doesn't justify the usage of a manual Future impl.

And actually for "waiting on something to be ready", I think it means we need a synchronization primitive. 
Here the async computation is not "owned" by the future, but we are just waiting for a notification.

Reply:
OK, fair point. Will refactor

Reply:
@xxchan How does it look now? I think you were right.